### PR TITLE
AVLN-0001 Integrate Pangolin DEX

### PR DESCRIPTION
**Pull request by kamilovsky01**
* Add helper view functions
* Add RefundRemainder modifier
* Integrate Pangolin DEX into the buyAsset method

### CODE REVIEW

File: contracts/SimpleAssetsExchange.sol [hunk 1]
```diff
@@ -1,24 +1,45 @@
...
- * @title SimpleAssetsExchange
+ * @title PangolinAssetsExchange
  * @dev Contract allows user to invest into an asset
- * It is a simple implementation that could be replace by a DEX or synthetic asset provider
+ * This implementation uses the Pangolin DEX
  */
-contract SimpleAssetsExchange is Ownable, IAssetsExchange {
-
+contract PangolinAssetsExchange is Ownable, IAssetsExchange {
   mapping(address => mapping(bytes32 =>uint256)) balance;
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:16]: The Pangolin DEX contract will act as a proxy to pangolin exchange - we don't need to keep tokens in this contract

File: contracts/SimpleAssetsExchange.sol [hunk 2]
```diff
@@ -33,24 +54,23 @@ contract SimpleAssetsExchange is Ownable, IAssetsExchange {
...
    * @dev _amount amount to be bought
   **/
-  function buyAsset(bytes32 _asset, uint256 _amount) payable override external {
+  function buyAsset(bytes32 _asset, uint256 _amount) payable override external RefundRemainder{
     uint256 amountIn = _amount * priceProvider.getPrice(_asset) / 1 ether;
     require(amountIn > 0, "Incorrect input amount");
     require(msg.value >= amountIn, "Not enough funds provided");
 
+    pangolinRouter.swapAVAXForExactTokens{ value: msg.value }(_amount, getPathForAVAXtoToken(_asset), address(this), block.timestamp + 100);
     balance[msg.sender][_asset] = balance[msg.sender][_asset] + _amount;
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:72]: Let's send tokens to msg.sender instead on keeping them on contract

File: contracts/SimpleAssetsExchange.sol [hunk 3]
```diff
@@ -65,6 +85,10 @@ contract SimpleAssetsExchange is Ownable, IAssetsExchange {
     uint256 amountOut = _amount * priceProvider.getPrice(_asset) / 1 ether;
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:85]: This won't work as the calculated amountOut may be different than the amount received from Pangolin

File: contracts/SimpleAssetsExchange.sol [hunk 4]
```diff
@@ -65,6 +85,10 @@ contract SimpleAssetsExchange is Ownable, IAssetsExchange {
...
     require(amountOut > 0, "Incorrect output amount");
 
+    IERC20 token = IERC20(assetAddress[_asset]);
+    token.approve(address(pangolinRouter), tokenInAmount);
+    pangolinRouter.swapExactTokensForAVAX(tokenInAmount, minAmountOut, getPathForTokenToAVAX(asset), address(this), block.timestamp + 100);
+
     balance[msg.sender][_asset] = balance[msg.sender][_asset] - _amount;
 
     payable(msg.sender).transfer(amountOut);
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:94]: This won't work as the calculated amountOut may be different than the amount received from Pangolin

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: contracts/SimpleAssetsExchange.sol [hunk 5]
```diff
@@ -33,24 +54,23 @@ contract SimpleAssetsExchange is Ownable, IAssetsExchange {
...
    * @dev _asset asset code
    * @dev _amount amount to be bought
   **/
-  function buyAsset(bytes32 _asset, uint256 _amount) payable override external {
+  function buyAsset(bytes32 _asset, uint256 _amount) payable override external RefundRemainder{
     uint256 amountIn = _amount * priceProvider.getPrice(_asset) / 1 ether;
     require(amountIn > 0, "Incorrect input amount");
     require(msg.value >= amountIn, "Not enough funds provided");
 
+    pangolinRouter.swapAVAXForExactTokens{ value: msg.value }(_amount, getPathForAVAXtoToken(_asset), address(this), block.timestamp + 100);
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:71]: What's the purpose of "block.timestamp+100" ?

Comment (kamilovsky01) [reply contracts/SimpleAssetsExchange.sol]: Nice catch. It was a leftover from JS scripts testing and is no longer required. Hence, I deleted it in the newest commit.

File: contracts/SimpleAssetsExchange.sol [hunk 6]
```diff
@@ -1,71 +1,70 @@
...
+   * Buys selected ERC20 token with AVAX using the Pangolin DEX
+   * Refunds unused AVAX to the msg.sender
+   * @dev _token ERC20 token's address
+   * @dev _amount amount of the ERC20 token to be bought
+   * TODO: Implement slippage % tolerance and add as a require check
   **/
-  function buyAsset(bytes32 _asset, uint256 _amount) payable override external {
-    uint256 amountIn = _amount * priceProvider.getPrice(_asset) / 1 ether;
+  function buyERC20Token(address _token, uint256 _amount) payable override external RefundRemainder{
     require(amountIn > 0, "Incorrect input amount");
```
Comment (jakub-wojciechowski) [inline contracts/SimpleAssetsExchange.sol:43]: Does it compile, amountIn is defined one line below?

**Comment by kamilovsky01** [comment]
This PR is now obsolete as there is no need to fork this repository after I was given write permissions to the main repository

**Pull request closed without merging**